README.md
# synthetic_control
A user friendly package to use the Synthetic Control Method


HISTORY.rst
History
=======

0.0.5 (2023-12-14)
-------------------
* Fix releases.

0.0.4 (2023-12-14)
-------------------
* Comment out setup.py long description.

0.0.3 (2023-12-14)
-------------------
* Fix pypi branch for github actions workflow.

0.0.2 (2023-12-14)
-------------------
* Fix tags.

0.0.1 (2023-12-14)
-------------------
* Initial release.


tests/test_model.py
# -*- coding: utf-8 -*-

""" Tests for synthetic_control.model. """

from datetime import datetime

import numpy as np
import pandas as pd
from pandas.testing import assert_frame_equal
from sklearn.linear_model import Ridge
from sklearn.utils.validation import check_is_fitted

from synthetic_control.model import SyntheticControl


class TestSyntheticControl:
    """Test the SyntheticControl class."""

    def get_test_data(self):
        """Create a test dataset."""
        index = pd.date_range(start="2006-01-01", end="2012-01-01", freq="YS")
        X = pd.DataFrame(
            {
                "Cat": [80, 84, 78, 90, 76, 72, 80],
                "Dog": [20, 16, 22, 10, 24, 28, 20],
                "Bird": [120, 80, 75, 182, 80, 60, 80],
                "Fish": [200, 400, 80, 129, 280, 320, 200],
            },
            index=index,
        )
        y = pd.Series(data=[100, 97, 96, 104, 92, 86, 92], index=index)
        return X, y

    def test_create(self):
        """Test the creation of a SyntheticControl object."""
        sc = SyntheticControl(treatment_start=datetime(2008, 1, 1))
        assert isinstance(sc, SyntheticControl)

    def test_setup_model(self):
        """Test the setup_model method."""
        sc = SyntheticControl(
            treatment_start=datetime(2008, 1, 1), treatment_end=datetime(2010, 1, 1)
        )
        sc._setup_model()
        assert isinstance(sc.model, Ridge)
        assert sc.treatment_start == datetime(2008, 1, 1)
        assert sc.treatment_end == datetime(2010, 1, 1)

    def test_get_treatment_phase(self):
        """Test the _create_treatment_phases method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(
            treatment_start=datetime(2008, 1, 1), treatment_end=datetime(2010, 1, 1)
        )
        treatment = sc._get_treatment_phase(X)
        assert np.allclose(treatment, [False] * 3 + [True] * 2 + [False] * 2)

    def test_get_treatment_phase_no_end(self):
        """Test the _create_treatment_phases method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(treatment_start=datetime(2008, 1, 1))
        treatment = sc._get_treatment_phase(X)
        assert np.allclose(treatment, [False] * 3 + [True] * 4)

    def test_fit(self):
        """Test the fit method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(treatment_start=datetime(2008, 1, 1))
        sc.fit(X, y)
        check_is_fitted(sc.model)

    def test_predict(self):
        """Test the predict method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(treatment_start=datetime(2008, 1, 1))
        sc.fit(X, y)
        y_synth = sc.predict(X)
        np.allclose(y_synth, [100.0, 97.1, 95.91, 106.6, 96.1, 93.9, 96.6], atol=0.1)

    def test_fit_predict(self):
        """Test the fit_predict method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(treatment_start=datetime(2008, 1, 1))
        y_synth = sc.fit_predict(X, y)
        np.allclose(y_synth, [100.0, 97.1, 95.91, 106.6, 96.1, 93.9, 96.6], atol=0.1)

    def test_get_confidence_interval(self):
        """Test the get_confidence_interval method."""
        X, y = self.get_test_data()
        sc = SyntheticControl(
            treatment_start=datetime(2008, 1, 1),
            ci_fraction=0.6,
            ci_percentiles=[5, 95],
        )
        y_ci = sc.get_confidence_interval(X, y)
        expected_df = pd.DataFrame(
            {
                5: [96.8, 81.2, 71.4, 51.4, 86.5, 70.7, 82.3],
                95: [118.1, 101.6, 100.1, 169.9, 121.1, 140.7, 98.2],
            },
            index=X.index,
        )
        assert_frame_equal(y_ci, expected_df, atol=0.1)


tests/analysis/test_compare.py
# -*- coding: utf-8 -*-

""" Tests for synthetic_control.analysis.compare. """


from datetime import datetime

import pandas as pd
import plotly.graph_objects as go

from synthetic_control.analysis import compare


class TestCompare:
    def tests_compare_to_synthetic_control(self):
        """Test the compare_to_synthetic_control function."""
        y = pd.Series([100, 90, 80])
        y_pred_ci = pd.DataFrame({5: [96, 81, 73], 95: [118, 101, 100]})
        fig = compare.compare_to_synthetic_control(
            y, y_pred_ci, datetime(2008, 1, 1), datetime(2010, 1, 1)
        )
        assert isinstance(fig, go.Figure)
        assert len(fig.data) == 4
        assert fig.data[0].name == "Treatment"
        assert fig.data[1].name == "Synthetic Control"
        assert fig.data[2].name == "90% confidence interval"
        assert fig.data[3].name == "90% confidence interval"

    def test_get_plot_data(self):
        """Test the get_plot_data function."""
        y = pd.Series([100, 90, 80])
        y_pred_ci = pd.DataFrame({5: [96, 81, 73], 95: [118, 101, 100]})
        data = compare.get_plot_data(y, y_pred_ci, "Treatment", show_impact=False)
        assert len(data) == 4
        assert all([isinstance(d, go.Scatter) for d in data])
        assert list(data[0].y) == [100, 90, 80]
        assert list(data[-1].y) == [118, 101, 100]

    def test_get_plot_data_impact(self):
        """Test the get_plot_data function."""
        y = pd.Series([100, 90, 80])
        y_pred_ci = pd.DataFrame({5: [96, 81, 73], 95: [118, 101, 100]})
        data = compare.get_plot_data(y, y_pred_ci, "Treatment", show_impact=True)
        assert len(data) == 4
        assert all([isinstance(d, go.Scatter) for d in data])
        assert list(data[0].y) == [0, 0, 0]
        assert list(data[-1].y) == [18, 11, 20]

    def test_get_baseline_prediction_with_median(self):
        """Test the get_baseline_prediction function with median."""
        y_pred_ci = pd.DataFrame({10: [10, 11, 12], 50: [20, 21, 22], 90: [26, 28, 28]})
        y_pred = compare.get_baseline_prediction(y_pred_ci)
        assert isinstance(y_pred, pd.Series)
        assert y_pred.equals(pd.Series([20, 21, 22]))

    def test_get_baseline_prediction_without_median(self):
        """Test the get_baseline_prediction function with median."""
        y_pred_ci = pd.DataFrame({10: [10, 11, 12], 45: [21, 21, 23], 90: [26, 28, 28]})
        y_pred = compare.get_baseline_prediction(y_pred_ci)
        assert isinstance(y_pred, pd.Series)
        assert y_pred.equals(pd.Series([19.0, 20.0, 21.0]))

    def test_add_confidence_interval(self):
        """Test the add_confidence_interval function."""
        y_pred_ci = pd.DataFrame(
            {5: [10, 11, 12], 25: [13, 15, 15], 75: [18, 19, 20], 95: [20, 21, 22]}
        )
        data = compare.add_confidence_interval(y_pred_ci, "red")
        assert len(data) == 4
        assert [d.name for d in data] == ["90% confidence interval"] * 4
        assert all([isinstance(d, go.Scatter) for d in data])

    def test_get_opacity_color(self):
        """Test the get_opacity_color function."""
        color = "rgb(140, 20, 14)"
        assert compare.get_opacity_color(color, opacity=0.5) == "rgba(140, 20, 14, 0.5)"

    def test_add_treatment_period(self):
        """Test the add_treatment_period function."""
        fig = go.Figure()
        fig = compare.add_treatment_period(
            fig, datetime(2008, 1, 1), datetime(2010, 1, 1)
        )
        assert len(fig.layout.shapes) == 2
        treatment_start = datetime.fromtimestamp(fig.layout.shapes[0].x0 / 1000)
        treatment_end = datetime.fromtimestamp(fig.layout.shapes[1].x0 / 1000)
        assert treatment_start == datetime(2008, 1, 1)
        assert treatment_end == datetime(2010, 1, 1)

    def test_get_plot_layout(self):
        """Test the get_plot_layout function."""
        layout = compare.get_plot_layout(y_axis="Test value", show_impact=False)
        assert isinstance(layout, go.Layout)
        assert layout.xaxis.title.text == "Date"
        assert layout.yaxis.title.text == "Test value"

    def test_get_plot_layout_impact(self):
        """Test the get_plot_layout function."""
        layout = compare.get_plot_layout(y_axis="Test value", show_impact=True)
        assert isinstance(layout, go.Layout)
        assert layout.xaxis.title.text == "Date"
        assert layout.yaxis.title.text == "Difference in Test value"


synthetic_control/model.py
# -*- coding: utf-8 -*-

""" A class to use the synthetic control method. """

import numpy as np
import pandas as pd
from sklearn.base import clone
from sklearn.linear_model import Ridge

from synthetic_control.analysis import compare


class SyntheticControl:
    """A model to use the synthetic control method.

    Parameters
    ----------
    treatment_start : datetime
        The start of the treatment period.
    treatment_end : datetime
        The end of the treatment period.
    treatment_name : str
        The name of the treatment group.
    ci_sample_size : int
        The number of times to generate predictions to build the confidence interval.
    ci_fraction : float
        The fraction of features to use when generating predictions to build the
        confidence interval. This is used to increase the variance across predictions.
        If the ci_fraction is 1.0, then all the predictions used to build the
        confidence interval are identical.
    ci_percentiles : tuple
        The percentiles to provide for the confidence interval.
    **kwargs
        Additional keyword arguments to pass to the fitting model.
    """

    CI_PERCENTILES = (5, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95)

    def __init__(
        self,
        treatment_start,
        treatment_end=None,
        treatment_name="Treatment",
        ci_sample_size=100,
        ci_fraction=0.1,
        ci_percentiles=CI_PERCENTILES,
        **kwargs,
    ):
        self.treatment_start = treatment_start
        self.treatment_end = treatment_end
        self.treatment_name = treatment_name
        self.ci_sample_size = ci_sample_size
        self.ci_fraction = ci_fraction
        self.ci_percentiles = ci_percentiles
        self.model = self._setup_model(**kwargs)

    def _setup_model(self, alpha=1.0, positive=True, fit_intercept=False):
        """Setup the model to build the synthetic control group.

        Parameters
        ----------
        alpha : float
            L2 regularization strength.
        positive : bool
            Force the linear coefficients to be positive. This is recommended for
            creating a reasonable synthetic control group.
        fit_intercept : bool
            Whether to fit the intercept for this model. It is recommended to se
            if to False for creating a reasonable synthetic control group.

        Returns
        -------
        sklearn.linear_model.Ridge
            A Ridge regression model used to build the synthetic control group.
        """
        return Ridge(alpha=alpha, positive=positive, fit_intercept=fit_intercept)

    def _get_treatment_phase(self, X):
        """Return the treatment phase. During the treatment itself, the model
        will not attempt to fit the synthetic control group to the treatment group.

        Parameters
        ----------
        X : pandas.DataFrame
            The data used to build the synthetic control group.

        Returns
        -------
        treatment : pandas.Series
            A boolean series indicating whether the observation occurs during the
            treatment.
        """
        treatment = X.index > self.treatment_start
        if self.treatment_end:
            treatment = treatment & (X.index <= self.treatment_end)
        return treatment

    def fit(self, X, y):
        """Fit the model to the data outside of the treatment period.

        Parameters
        ----------
        X : pandas.DataFrame
            The data used to build the synthetic control group.
        y : pandas.Series
            The treatment group to match outside of the treatment period.

        Returns
        -------
        self : SyntheticControl
            The SyntheticControl with the fitted model.
        """
        fitting_period = ~self._get_treatment_phase(X)
        X_train = X.loc[fitting_period]
        y_train = y.loc[fitting_period]
        self.model.fit(X_train, y_train)
        return self

    def predict(self, X):
        """Generate the predictions for the synthetic control group.

        Parameters
        ----------
        X : pandas.DataFrame
            The data used to build the synthetic control group.

        Returns
        -------
        y_pred : pandas.Series
            The predictions for the synthetic control group.
        """
        return self.model.predict(X)

    def fit_predict(self, X, y):
        """Fit the model to the data outside of the treatment period and generate
        the predictions for the synthetic control group.

        Parameters
        ----------
        X : pandas.DataFrame
            The data used to build the synthetic control group.
        y : pandas.Series
            The treatment group to match outside of the treatment period.

        Returns
        -------
        y_pred : pandas.Series
            The predictions for the synthetic control group.
        """
        self.fit(X, y)
        return self.model.predict(X)

    def get_confidence_interval(self, X, y):
        """Return the confidence interval on the predictions for the synthetic
        control group. In order to generate a range of possible predictions, the
        pre-treatment fitting model is trained on a subset (ci_fraction) of the
        features.

        Parameters
        ----------
        X : pandas.DataFrame
            The data used to build the synthetic control group.
        y : pandas.Series
            The treatment group to match outside of the treatment period.

        Returns
        -------
        y_pred_ci : pandas.DataFrame
            The quantile predictions for the synthetic control group.
        """
        preds = []
        for i in range(self.ci_sample_size):
            X_iter = X.T.sample(frac=self.ci_fraction, random_state=i, replace=False).T
            model = clone(self.model)
            fitting_period = ~self._get_treatment_phase(X)
            X_train = X_iter.loc[fitting_period]
            y_train = y.loc[fitting_period]
            model.fit(X_train, y_train)
            y_pred = model.predict(X_iter)
            preds.append(y_pred)
        ci = {q: np.percentile(preds, q=q, axis=0) for q in self.ci_percentiles}
        return pd.DataFrame(ci, index=X.index)

    def compare(self, y, y_pred_ci, y_axis="Value"):
        return compare.compare_to_synthetic_control(
            y,
            y_pred_ci,
            treatment_start=self.treatment_start,
            treatment_end=self.treatment_end,
            treatment_name=self.treatment_name,
            y_axis=y_axis,
            show_impact=False,
        )

    def impact(self, y, y_pred_ci, y_axis="Value"):
        return compare.compare_to_synthetic_control(
            y,
            y_pred_ci,
            treatment_start=self.treatment_start,
            treatment_end=self.treatment_end,
            treatment_name=self.treatment_name,
            y_axis=y_axis,
            show_impact=True,
        )


synthetic_control/analysis.py
# -*- coding: utf-8 -*-

""" Functions to provide analysis for the synthetic control method. """

from datetime import datetime

import numpy as np
import plotly.graph_objects as go
from plotly.colors import DEFAULT_PLOTLY_COLORS


def display_synthetic_control(
    y,
    y_pred_ci,
    treatment_start,
    treatment_end=None,
    treatment_name="Treatment",
    y_axis="Value",
):
    """ """
    data = get_plot_data(y, y_pred_ci, treatment_name)
    layout = get_plot_layout(y_axis)
    fig = go.Figure(data=data, layout=layout)
    fig = add_treatment_period(fig, treatment_start, treatment_end)
    return fig


def get_plot_data(y, y_pred_ci, treatment_name):
    """ """
    treatment_color = DEFAULT_PLOTLY_COLORS[0]
    control_color = DEFAULT_PLOTLY_COLORS[1]
    y_pred = get_base_prediction(y_pred_ci)
    data = [
        go.Scatter(
            x=y.index,
            y=y,
            name=treatment_name,
            line_color=treatment_color,
            line_width=5,
        ),
        go.Scatter(
            x=y.index,
            y=y_pred,
            name="Synthetic Control",
            line_color=control_color,
            line_width=5,
        ),
    ]
    data = add_confidence_interval(data, y_pred_ci, control_color)
    return data


def get_base_prediction(y_pred_ci):
    """ """
    if 50 in y_pred_ci:
        y_pred = y_pred_ci[50]
    else:
        y_pred = y_pred_ci.mean(axis=1)
    return y_pred


def add_confidence_interval(data, y_pred_ci, color):
    """ """
    min_ci = y_pred_ci.columns.min()
    max_ci = y_pred_ci.columns.max()
    ci_range = max_ci - min_ci
    color = get_opacity_color(color)
    columns = y_pred_ci.columns.sort_values(
        key=lambda x: np.abs(50 - x), ascending=False
    )
    for i, col in enumerate(columns):
        data.append(
            go.Scatter(
                x=y_pred_ci.index,
                y=y_pred_ci[col],
                name=f"{ci_range}% confidence interval",
                fill="tonexty",
                line_color=color,
                fillcolor=color,
                showlegend=i == 0,
                legendgroup="CI",
            )
        )
    return data


def get_opacity_color(color):
    return color.replace("rgb", "rgba").replace(")", ", 0.1)")


def add_treatment_period(fig, treatment_start, treatment_end):
    """ """
    fig.add_vline(
        x=treatment_start.timestamp() * 1000,
        line_color="black",
        line_dash="dash",
        annotation={"text": "Treatment Start", "xanchor": "center", "y": 1.1},
    )
    if treatment_end:
        fig.add_vline(
            x=treatment_end.timestamp() * 1000,
            line_color="black",
            line_dash="dash",
            annotation={"text": "Treatment End", "xanchor": "center", "y": 1.1},
        )
    return fig


def get_plot_layout(y_axis):
    """ """
    layout = go.Layout(xaxis={"title": "Date"}, yaxis={"title": y_axis})
    return layout


synthetic_control/analysis/compare.py
# -*- coding: utf-8 -*-

""" Functions to provide analysis for the synthetic control method. """

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.colors import DEFAULT_PLOTLY_COLORS


def compare_to_synthetic_control(
    y,
    y_pred_ci,
    treatment_start,
    treatment_end=None,
    treatment_name="Treatment",
    y_axis="Value",
    show_impact=False,
):
    """Display a comparison between the treatment group and the synthetic control group.

    Parameters
    ----------
    y : pandas.Series
        The treatment group.
    y_pred_ci : pandas.DataFrame
        The synthetic control group.
    treatment_start : datetime
        The start of the treatment period.
    treatment_end : datetime
        The end of the treatment period.
    treatment_name : str
        The name of the treatment group.
    y_axis : str
        The label for the y-axis.
    show_impact : bool
        Whether to show the impact of the treatment on the treatment group instead of
        the absolute values.

    Returns
    -------
    fig : plotly.graph_objects.Figure
        The plotly figure showing the comparison between the treatment group and the
        synthetic control group.
    """
    data = get_plot_data(y, y_pred_ci, treatment_name, show_impact)
    layout = get_plot_layout(y_axis, show_impact)
    fig = go.Figure(data=data, layout=layout)
    fig = add_treatment_period(fig, treatment_start, treatment_end)
    return fig


def get_plot_data(y, y_pred_ci, treatment_name, show_impact):
    """Create the plotly data for the comparison plot between the treatment group
    and the synthetic control group, including the confidence intervals.

    Parameters
    ----------
    y : pandas.Series
        The actual value of the metric for the treatment group.
    y_pred_ci : pandas.DataFrame
        The predicted values for each percentile for the synthetic control group.
    treatment_name : str
        The name of the treatment group.
    show_impact : bool
        Whether to show the impact of the treatment on the treatment group instead of
        the absolute values.

    Returns
    -------
    data : list
        The list of plotly data objects to display the comparison between the
        treatment group and the synthetic control group.
    """
    treatment_color = DEFAULT_PLOTLY_COLORS[0]
    control_color = DEFAULT_PLOTLY_COLORS[1]
    y_pred = get_baseline_prediction(y_pred_ci)
    if show_impact:
        y_pred = (y_pred - y).copy()
        y_pred_ci = (y_pred_ci.subtract(y, axis=0)).copy()
        y = pd.Series(0, index=y.index)
    data = [
        go.Scatter(
            x=y.index,
            y=y,
            name=treatment_name,
            line_color=treatment_color,
            line_width=5,
        ),
        go.Scatter(
            x=y.index,
            y=y_pred,
            name="Synthetic Control",
            line_color=control_color,
            line_width=5,
        ),
    ]
    data += add_confidence_interval(y_pred_ci, control_color)
    return data


def get_baseline_prediction(y_pred_ci):
    """Return the baseline prediction for the synthetic control group. The baseline
    prediction is the prediction for the 50th percentile if available, or the average
    of the predictions for all percentiles otherwise.

    Parameters
    ----------
    y_pred_ci : pandas.DataFrame
        The predicted values for each percentile for the synthetic control group.

    Returns
    -------
    y_pred : pandas.Series
        The baseline prediction for the synthetic control group.
    """
    if 50 in y_pred_ci:
        y_pred = y_pred_ci[50]
    else:
        y_pred = y_pred_ci.mean(axis=1)
    return y_pred


def add_confidence_interval(y_pred_ci, color):
    """Add the confidence interval to the plotly data.

    Parameters
    ----------
    y_pred_ci : pandas.DataFrame
        The predicted values for each percentile for the synthetic control group.
    color : str
        The color to use for the synthetic control group.

    Returns
    -------
    ci_data : list
        The list of plotly data objects to display the confidence interval.
    """
    min_ci = y_pred_ci.columns.min()
    max_ci = y_pred_ci.columns.max()
    ci_range = max_ci - min_ci
    color = get_opacity_color(color)
    columns = y_pred_ci.columns.sort_values(
        key=lambda x: np.abs(50 - x), ascending=False
    )
    ci_data = []
    for i, col in enumerate(columns):
        ci_data.append(
            go.Scatter(
                x=y_pred_ci.index,
                y=y_pred_ci[col],
                name=f"{ci_range}% confidence interval",
                fill="tonexty" if i > 0 else "none",
                line_color=color,
                fillcolor=color,
                showlegend=i == 1,
                legendgroup="CI",
            )
        )
    return ci_data


def get_opacity_color(color, opacity=0.1):
    """Return the color with the specified opacity.

    Parameters
    ----------
    color : str
        A color in rgb string format.
    opacity : float
        The opacity to apply to the color.

    Returns
    -------
    color : str
        The color with the specified opacity.
    """
    return color.replace("rgb", "rgba").replace(")", f", {opacity})")


def add_treatment_period(fig, treatment_start, treatment_end):
    """Add vertical lines to the plotly figure to delimitate the treatment period.

    Parameters
    ----------
    fig : plotly.graph_objects.Figure
        The plotly figure to add the vertical lines to.
    treatment_start : datetime
        The start of the treatment period.
    treatment_end : datetime
        The end of the treatment period.

    Returns
    -------
    fig : plotly.graph_objects.Figure
        The plotly figure with the vertical lines added.
    """
    fig.add_vline(
        x=treatment_start.timestamp() * 1000,
        line_color="black",
        line_dash="dash",
        annotation={"text": "Treatment Start", "xanchor": "center", "y": 1.1},
    )
    if treatment_end:
        fig.add_vline(
            x=treatment_end.timestamp() * 1000,
            line_color="black",
            line_dash="dash",
            annotation={"text": "Treatment End", "xanchor": "center", "y": 1.1},
        )
    return fig


def get_plot_layout(y_axis, show_impact):
    """Return the plotly layout for the comparison plot between the treatment group
    and the synthetic control group.

    Parameters
    ----------
    y_axis : str
        The label for the y-axis.
    show_impact : bool
        Whether to show the impact of the treatment on the treatment group instead of
        the absolute values.

    Returns
    -------
    layout : plotly.graph_objects.Layout
        The plotly layout for the comparison plot between the treatment group and the
        synthetic control group.
    """
    y_title = f"Difference in {y_axis}" if show_impact else y_axis
    layout = go.Layout(
        xaxis={"title": "Date"}, yaxis={"title": y_title}, template="none"
    )
    return layout
